New+ Support for variables in template filenames (35287)

cgaarden opened the pull request:
## Summary of the Pull Request
### Added support for variables in template filenames
Variables supported includes
1. Date and time component variables, such as `$YYYY`, `$YY`, `$MM`, `$DD`, - using the same syntax and implementation as in PowerRename
2. `$PARENT_FOLDER_NAME` - will replace variable with the name of the parent folder
3. `%environment-variable%` - will replace variable with the value of the corresponding environment variable

Example:
Template filename:
`$YYYY-$MM-%DD, $hh - $PARENT_FOLDER_NAME by %USERNAME%`

Filename on create becomes:
`2025-01-25, 02 - PowerShell project by cgaarden`

## PR Checklist
- [x] **Closes:** #35287
-  [x] **Also closes:** 
-  [x] **35287
-  [x] **35984
-  [x] **36615
-  [x] **36425
- [x] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [n/a] **Tests:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [n/a] **Dev docs:** Added/updated
- [n/a] **New binaries:** Added on the required places
   - [n/a] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [n/a] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [n/a] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [n/a] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [x] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #5285
https://github.com/MicrosoftDocs/windows-dev-docs/pull/5285

## Detailed Description of the Pull Request / Additional comments
### SHFileOperation API related
Still using SHFileOperation API - may change this in a future PR as the current implementation requires iterating through files multiple times (not really that impactful unless a template folder contains 100s+ of files/subfolders
Fixed an issue on ARM64 when depending on SHFileOperation to provide filename remapping  

### Minor refactoring
To help with support the variable support

### Minor exception handling logic
To help get information on where New+ menu isn't being displayed

## Validation Steps Performed
1. Windows 10 clean install, with lots of different template permutations
2. Windows 11 clean and update install, with lots of different template permutations
3. Windows 11 ARM64 and x64, with lots of different template permutations
5. Template folders with subitems with multiple variables within filenames
6. Template items starting with digits and spaces, e.g. 13. $YYYY-$MM-$DD PTA meeting minutes and actions
7. Rename mode is entered as expected for files and folder based templates, also when creating template with same name multiple times (x) in Arabic script
8. Validated that GPO ConfigureEnabledUtilityNewPlus, NewPlusHideTemplateFilenameExtension, NewPlusReplaceVariablesInTemplateFilenames are working as intended

### Example template folder + corresponding context menu
![image](https://github.com/user-attachments/assets/38818701-355c-4091-884c-039ebb3c85d0)


### Example of project including subitems
![image](https://github.com/user-attachments/assets/27c3f2c3-025b-42a4-a566-3b9009ebeaf9)


### New+ settings 
(debug)
![image](https://github.com/user-attachments/assets/be1ca7f1-b148-4cd3-aab3-60cc27b35e3a)


### New+ related GPO policies
![image](https://github.com/user-attachments/assets/4ca47265-9da2-4233-9384-e407adfd4b8d)

crutkas commented:
Has conflicts

cgaarden commented:
> Has conflicts

Recent conflicts with GPO settings are now resolved again

jaimecbernardo commented on src/gpo/assets/en-US/PowerToys.adml:
```diff
@@ -289,7 +297,6 @@ If you don't configure this policy, the user takes control over the setting and
       <string id="MwbPolicyDefinedIpMappingRules">Predefined IP Address mapping rules</string>
       <string id="NewPlusHideTemplateFilenameExtension">Hide template filename extension</string>
       <string id="AllowDiagnosticData">Allow sending diagnostic data</string>
-      <string id="ConfigureRunAtStartup">Configure the run at startup setting</string>
```
Why is this string removed? I suspect a bad merge.

cgaarden replied:
Argh. Thanks for spotting that. Yes bad merge.

jaimecbernardo commented on src/settings-ui/Settings.UI.Library/NewPlusProperties.cs:
```diff
@@ -18,6 +18,7 @@ public NewPlusProperties()
             HideFileExtension = new BoolProperty(true);
             HideStartingDigits = new BoolProperty(true);
             TemplateLocation = new StringProperty(Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Microsoft", "PowerToys", "NewPlus", "Templates"));
+            ReplaceVariables = new BoolProperty(true);
```
Perhaps default should be false? Since this new feature might break existing behavior in an unexpected way for some users right after updating PowerToys.

cgaarden replied:
Changed the default to false (needed to be changed here and in settings.cpp)

cgaarden replied:
Updated documentation as well

jaimecbernardo commented:
/azp run

jaimecbernardo reviewed (changes requested):
I meant request change instead of approve, sorry 😅

cgaarden commented:
> Hi @cgaarden , Thank you for opening the PR.
> 
> I think the code looks good and I like the approach. I tried to use the docs and not everything worked for me:
> 
> "$YYYY-$MM-%DD, $hh $mm $ss - $PARENT_FOLDER_NAME by %USERNAME%" I tried this and saw that there's a error in `%DD` instead of `$DD` , but still couldn't get $PARENT_FOLDER_NAME to work for me. I left some smaller comments a bit before writing this review. Thank you :)

Fixed %DD in the documentation.  Thank you!

Regarding $PARENT_FOLDER_NAME
$PARENT_FOLDER_NAME only works in subfolders, e.g.
Templates\Key project\$PARENT_FOLDER_NAME intro.md

will expand to
<destination>\Key project\Key project intro.md

and doesn't work at the "root" of the template folder, e.g.
Templates\$PARENT_FOLDER_NAME intro.md

cgaarden commented:
Thank you @jaimecbernardo for review and feedback!

I've updated the PR and also the documentation PR. 

Thank you!

jaimecbernardo commented:
/azp run

jaimecbernardo commented:
/azp run

jaimecbernardo reviewed (changes requested):
Hi Christian. Gave it another try, but seems like the new setting doesn't do anything. Variables will be replaced whether that setting is off or on. Can you please take a look? Thank you!

cgaarden commented:
> Hi Christian. Gave it another try, but seems like the new setting doesn't do anything. Variables will be replaced whether that setting is off or on. Can you please take a look? Thank you!

Thank you Jaime for having yet another look. 

I fixed the issue and fixed a similar issue that I completely missed for optionally hiding file extensions. 

Thank you for the feedback and patience with my mistakes here.

jaimecbernardo commented:
/azp run
